File: Embarcadero/Rio/CPP/NotesResource/NotesStorageU.cpp
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

#pragma hdrstop
#pragma package(smart_init)

#include "NotesStorageU.h"
#include <memory>

// ---------------------------------------------------------------------------
String __fastcall TNotesStorage::DecodeMultilineText(const String & AText) {
	std::auto_ptr<TStringBuilder> lBuilder(new TStringBuilder());
	String _return = "";
	int i = 1;
	while(i <= AText.Length()) {
		WideChar S = AText[i];
		if((S == '\\') && (i+1 <= AText.Length())) {
			switch(AText[i+1]) {
				case 'n' :
					i++;
					S = static_cast<WideChar>(10);
					break;
				case 'r' :
					i++;
					S = static_cast<WideChar>(13);
					break;
				case '\\' :
					i++;
					S = '\\';
					break;
			}
		}
		lBuilder->Append(S);
		i++;
	}
	_return = lBuilder->ToString();
	return _return;
}
// ---------------------------------------------------------------------------
String __fastcall TNotesStorage::EncodeMultilineText(const String & AText) {
	std::auto_ptr<TStringBuilder> lBuilder(new TStringBuilder());
	String _return = "";
	for(int i = 1; i <= AText.Length(); i++)
	{
		WideChar S = AText[i];
		switch(S) {
			case 10 :
				lBuilder->Append("\\n");
				break;
			case 13 :
				lBuilder->Append("\\r");
				break;
			case '\\' :
				lBuilder->Append("\\\\");
				break;
			default:
				lBuilder->Append(S);
				break;
		}
	}
	_return = lBuilder->ToString();
	return _return;
}
// ---------------------------------------------------------------------------
void __fastcall TNotesStorage::ReadNote(const String & AID, TNote *ANote) {
	ANote->Title = FIniFile->ReadString(FUserID, AID + ".title", "");
	ANote->Content = DecodeMultilineText(FIniFile->ReadString(FUserID, AID + ".text", ""));
	ANote->ID = AID;
}
// ---------------------------------------------------------------------------
void __fastcall TNotesStorage::ReadIDs(std::vector<String> * AList) {
	TStrings * lSections = new TStringList();
	FIniFile->ReadSection(FUserID, lSections);
	for(int i = 0; i < lSections->Count; i++)
	{
		String lKey = lSections->operator [](i);
		if(EndsStr(".title", lKey)) {
			String lId = lKey.SubString(0, lKey.Length() - String(".title").Length());
			AList->push_back(lId);
        }
	}
}
// ---------------------------------------------------------------------------
std::vector<String> * __fastcall TNotesStorage::ReadIDs(void) {
	std::vector<String> * _return = new std::vector<String>();
	ReadIDs(_return);
	return _return;
}
// ---------------------------------------------------------------------------
__fastcall TNotesStorage::TNotesStorage(const String & ADirectory,
	const String & AUserID) {
	FUserID = AUserID;
	String lPath = IncludeTrailingPathDelimiter(ExpandFileName(ADirectory)) + "notes.ini";
	FIniFile = new TIniFile(lPath);
}
// ---------------------------------------------------------------------------
__fastcall TNotesStorage::~TNotesStorage(void) {
	FreeAndNil(FIniFile);
}
// ---------------------------------------------------------------------------
std::vector<TNote*> * __fastcall TNotesStorage::GetNotes(void) {
	std::vector<TNote*> * notes = new std::vector<TNote*>();

	std::vector<String> *ids = ReadIDs();
	for(std::vector<String>::iterator it = ids->begin(); it != ids->end(); it++)
	{
		TNote * note = new TNote();
		ReadNote(*it, note);
		notes->push_back(note);
	}
	 return notes;
}
// ---------------------------------------------------------------------------
bool __fastcall TNotesStorage::GetNote(const String & AID, TNote * ANote) {
	bool _return = NoteIDExists(AID);
	if(_return) {
		ReadNote(AID, ANote);
	}
	return _return;
}
// ---------------------------------------------------------------------------
void __fastcall TNotesStorage::UpdateNote(const String & AID,
	const TNote * ANote) {
	if(!NoteIDExists(AID)) {
		throw ENoteNotFound(String().printf(L"%s", AID.c_str()));
	}
	FIniFile->WriteString(FUserID, AID + ".title", EncodeMultilineText(ANote->Title));
	FIniFile->WriteString(FUserID, AID + ".text", EncodeMultilineText(ANote->Content));
}

String __fastcall TNotesStorage::NextId(void)
{
	std::vector<String> * lList = new std::vector<String>();
	ReadIDs(lList);
	int i = lList->size();
	for(std::vector<String>::iterator it = lList->begin(); it != lList->end(); it++)
	{
		if(ContainsStr(*it, IntToStr(i))) {
			break;
		}
		i++;
	}
	delete lList;
	return IntToStr(i);
}

// ---------------------------------------------------------------------------
void __fastcall TNotesStorage::AddNote(const TNote & ANote, String &AID) {
	if(ANote.Title.IsEmpty()) {
		throw new ENoteMissingTitle("Note title required");
	}
	if(NoteTitleExists(ANote.Title)) {
		throw ENoteDuplicate(String().sprintf(L"%s already exists", ANote.Title.c_str()));
    }
	AID = this->NextId();
	FIniFile->WriteString(FUserID, AID + ".title", ANote.Title);
	FIniFile->WriteString(FUserID, AID + ".text", EncodeMultilineText(ANote.Content));
}
// ---------------------------------------------------------------------------
bool __fastcall TNotesStorage::DeleteNote(const String & AID) {
	bool _return = NoteIDExists(AID);
	if(_return) {
        FIniFile->DeleteKey(FUserID, AID + ".title");
	}
	return _return;
}
// ---------------------------------------------------------------------------
bool __fastcall TNotesStorage::NoteTitleExists(const String ATitle) {
	bool _return = false;
	std::vector<TNote*> * notes = this->GetNotes();
	for(std::vector<TNote*>::iterator it = notes->begin(); it != notes->end(); it++)
	{
		if((*it)->Title == ATitle) {
			_return = true;
			break;
		}
	}
	return _return;
}
// ---------------------------------------------------------------------------
bool __fastcall TNotesStorage::FindNote(const String &ATitle, TNote * ANote) {
	bool _return = false;
	std::vector<TNote*> * notes = this->GetNotes();
	for(std::vector<TNote*>::iterator it = notes->begin(); it != notes->end(); it++)
	{
		if((*it)->Title == ATitle) {
			ReadNote((*it)->ID, ANote);
			_return = true;
			break;
		}
	}
	return _return;
}
// ---------------------------------------------------------------------------
bool __fastcall TNotesStorage::NoteIDExists(const String &AID) {
	return FIniFile->ValueExists(FUserID, AID + ".title");
}
// ---------------------------------------------------------------------------

File: Embarcadero/Rio/Object Pascal/NotesResource/NotesStorageU.pas
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

unit NotesStorageU;

interface

uses System.Classes, System.SysUtils,  System.IniFiles, System.Generics.Collections,
  NoteTypesU;

type

  TNotesStorage = class
  private
    FIniFile: TIniFile;
    FUserID: string;
    function DecodeMultilineText(const AText: string): string;
    function EncodeMultilineText(const AText: string): string;
    procedure ReadNote(const AID: string; out ANote: TNote);
    procedure ReadIDs(const AList: TList<string>); overload;
    function ReadIDs: TArray<string>; overload;
  public
    constructor Create(const ADirectory: string; const AUserID: string);
    destructor Destroy; override;
    function GetNotes: TArray<TNote>;
    function GetNote(const AID: string; out ANote: TNote): Boolean;
    procedure UpdateNote(const AID: string; const ANote: TNote);
    procedure AddNote(const ANote: TNote; out AID: string);
    function DeleteNote(const AID: string): Boolean;
    function NoteTitleExists(const ATitle: string): Boolean;
    function FindNote(const ATitle: string; out ANote: TNote): Boolean;
    function NoteIDExists(const AID: string): Boolean;
  end;

  ENoteError = class(Exception);
  ENoteNotFound = class(ENoteError);
  ENoteDuplicate = class(ENoteError);
  ENoteMissingTitle = class(ENoteError);

implementation

{ TNotesManager }

procedure TNotesStorage.AddNote(const ANote: TNote; out AID: string);

  function NextID: string;
  var
    LList: TList<string>;
    I: Integer;
  begin
    LList := TList<string>.Create;
    try
      ReadIDs(LList);
      I := LList.Count;
      while LList.Contains(IntToStr(I)) do
        Inc(I);
      Result := IntToStr(I);
    finally
      LList.Free;
    end;
  end;

begin
  if ANote.Title = '' then
    raise ENoteMissingTitle.Create('Note title required');
  if NoteTitleExists(ANote.Title) then
    raise ENoteDuplicate.CreateFmt('"%s" already exists', [ANote.Title]);
  AID := NextID;
  FIniFile.WriteString(FUserID, AID + '.title', ANote.Title);
  FIniFile.WriteString(FUserID, AID + '.text', EncodeMultilineText( ANote.Content));
end;

constructor TNotesStorage.Create(const ADirectory: string; const AUserID: string);
var
  LPath: string;
begin
  FUserID := AUserID;
  LPath := IncludeTrailingPathDelimiter(ExpandFileName(ADirectory)) + 'notes.ini';
  FIniFile := TIniFile.Create(LPath);
end;

function TNotesStorage.NoteIDExists(const AID: string): Boolean;
begin
  Result := FIniFile.ValueExists(FUserID, AID + '.title');
end;

function TNotesStorage.NoteTitleExists(const ATitle: string): Boolean;
var
  LNote: TNote;
begin
  Result := False;
  for LNote in GetNotes do
    if LNote.Title = ATitle then
    begin
      Result := True;
      break;
    end;
end;

function TNotesStorage.FindNote(const ATitle: string; out ANote: TNote): Boolean;
var
  LNote: TNote;
begin
  Result := False;
  for LNote in GetNotes do
    if LNote.Title = ATitle then
    begin
      Result := True;
      ReadNote(LNote.ID, ANote);
      break;
    end;
end;

function TNotesStorage.DeleteNote(const AID: string): Boolean;
begin
  Result := NoteIDExists(AID);
  if Result then
    FIniFile.DeleteKey(FUserID, AID + '.title');
end;

destructor TNotesStorage.Destroy;
begin
  FIniFile.Free;
  inherited;
end;

function TNotesStorage.EncodeMultilineText(const AText: string): string;
var
  LBuilder: TStringBuilder;
  S: Char;
begin
  LBuilder := TStringBuilder.Create;
  try
    for S in AText do
    begin
      case S of
      #10:
        LBuilder.Append('\n');
      #13:
        LBuilder.Append('\r');
      '\':
        LBuilder.Append('\\');
      else
        LBuilder.Append(S);
      end;
    end;
    Result := LBuilder.ToString;
  finally
    LBuilder.Free;
  end;
end;

function TNotesStorage.DecodeMultilineText(const AText: string): string;
var
  LBuilder: TStringBuilder;
  I: Integer;
  S: Char;
begin
  LBuilder := TStringBuilder.Create;
  try
    I := 0;
    while I < AText.Length do
    begin
      S := AText.Chars[I];
      if (S = '\') and (I+1 < AText.Length) then
        case AText.Chars[I+1] of
          'n':
            begin
              Inc(I);
              S := #10;
            end;
          'r':
            begin
              Inc(I);
              S := #13;
            end;
          '\':
            begin
              Inc(I);
              S := '\';
            end
        end;
      LBuilder.Append(S);
      Inc(I);
    end;
    Result := LBuilder.ToString;
  finally
    LBuilder.Free;
  end;
end;

function TNotesStorage.GetNote(const AID: string; out ANote: TNote): Boolean;
begin
  Result := NoteIDExists(AID);
  if Result then
    ReadNote(AID, ANote);
end;

procedure TNotesStorage.ReadNote(const AID: string; out ANote: TNote);
var
  LText: string;
  LTitle: string;
begin
  LTitle := FIniFile.ReadString(FUserID, AID + '.title', '');
  LText := DecodeMultilineText(FIniFile.ReadString(FUserID, AID + '.text', ''));
  ANote := TNote.CreateWithID(LTitle, LText, AID);
end;

function TNotesStorage.GetNotes: TArray<TNote>;
var
  LList: TList<TNote>;
  LNote: TNote;
  LSection: TStrings;
  LID: string;
begin
  LSection := nil;
  LList := nil;
  try
    LSection := TStringList.Create;
    LList := TList<TNote>.Create;
    for LID in ReadIDs do
    begin
      ReadNote(LID, LNote);
      LList.Add(LNote);
    end;
    Result := LList.ToArray;
  finally
    LList.Free;
    LSection.Free;
  end;
end;


function TNotesStorage.ReadIDs: TArray<string>;
var
  LList: TList<string>;
begin
  LList := TList<string>.Create;
  try
    ReadIDs(LList);
    Result := LList.ToArray;
  finally
    LList.Free;
  end;
end;

procedure TNotesStorage.ReadIDs(const AList: TList<string>);
var
  LSection: TStrings;
  I: Integer;
  LID: string;
  LKey: string;
begin
  LSection := nil;
  try
    LSection := TStringList.Create;
    FIniFile.ReadSection(FUserID, LSection);
    for I := 0 to LSection.Count - 1 do
    begin
      LKey := LSection[I];
      if LKey.EndsWith('.title') then
      begin
        LID := LKey.Substring(0, Length(LKey) - Length('.title'));
        Assert(not AList.Contains(LID));
        Assert(LID <> '');
        AList.Add(LID);
      end;
    end;
  finally
    LSection.Free;
  end;
end;

procedure TNotesStorage.UpdateNote(const AID: string; const ANote: TNote);
begin
  if not NoteIDExists(AID) then
    raise ENoteNotFound.CreateFmt('"%s" not found', [AID]);
  FIniFile.WriteString(FUserID, AID + '.title', EncodeMultilineText(ANote.Title));
  FIniFile.WriteString(FUserID, AID + '.text', EncodeMultilineText(ANote.Content));
end;


end.
